fish-rust/src/builtins/functions.rs

```diff
@@ -0,0 +1,413 @@
+        }
+        if let Some(props) = function::get_props_autoload(args[0], parser) {
+            if let Some(cpf) = &props.copy_definition_file {
+                streams.out.append(sprintf!("%ls\n", cpf));
+            } else if let Some(df) = &props.definition_file {
+                streams.out.append(sprintf!("%ls\n", df));
+            } else {
+                streams.out.append(sprintf!("%ls\n", "stdin"));
+            }
+            if opts.verbose {
+                if props.copy_definition_file.is_some() {
+                    if let Some(df) = &props.definition_file {
+                        streams.out.append(sprintf!("%ls\n", df));
+                    } else {
+                        streams.out.append(sprintf!("%ls\n", "stdin"));
+                    }
+                } else {
+                    streams.out.append(sprintf!(
+                        "%ls\n",
+                        if props.is_autoload.load() {
+                            "autoloaded"
+                        } else {
+                            "n/a"
+                        }
+                    ));
+                }
+                streams.out.append(sprintf!(
+                    "%d\n%ls\n",
+                    props.definition_lineno(),
+                    if props.shadow_scope {
+                        "scope-shadowing"
+                    } else {
+                        "no-scope-shadowing"
+                    }
+                ));
+                if !props.description.is_empty() {
+                    streams.out.append(sprintf!(
+                        "%ls\n",
+                        escape_string(
+                            &props.description,
+                            EscapeStringStyle::Script(
+                                EscapeFlags::NO_PRINTABLES | EscapeFlags::NO_QUOTED
+                            )
+                        ),
+                    ));
+                }
+            }
+        } else if opts.verbose {
+            streams.out.append(L!("n/a\nn/a\n0n/a\nn/a\n"));
+        } else {
+            streams.out.append(L!("n/a\n"));
+        }
+        // Historical - this never failed?
+        return STATUS_CMD_OK;
+    }
+
+    if opts.handlers {
+        if !opts.handlers_type.is_empty()
+            && !event::EVENT_FILTER_NAMES.contains(&opts.handlers_type.as_utfstr())
+        {
+            streams.err.append(wgettext_fmt!(
+                "%ls: Expected generic | variable | signal | exit | job-id for --handlers-type\n",
+                cmd
+            ));
+            return STATUS_INVALID_ARGS;
+        }
+        event::print(streams, &opts.handlers_type);
+        return STATUS_CMD_OK;
+    }
+
+    if opts.query && args.is_empty() {
+        return STATUS_CMD_ERROR;
+    }
+
+    if opts.list || args.is_empty() {
+        let mut names = function::get_names(opts.show_hidden);
+        names.sort();
+        if !streams.out_is_redirected && unsafe { isatty(STDOUT_FILENO) == 1 } {
```

Comment:
We've had this a few times, a helper method might be nice to avoid multiple "unsafe" blocks.

Reply:
Once IO is ported, we can bump our MSRV to 1.70 so we can use [`is_terminal`](https://doc.rust-lang.org/stable/std/io/trait.IsTerminal.html). Unless that has some behavior-difference I am not aware of, have not seen the internals of it

Reply:
The idea of increasing a dependency requirement to replace a single line with a slightly shorter single line sounds unappealing to me.

Especially when that line does:

https://github.com/rust-lang/rust/blob/4d7a80d48697171ed151c4667b74fc7d784c1836/library/std/src/sys/unix/io.rs#L79-L82

The *exact same thing we already do* - excluding the "out_is_redirected" check.

I would simply add a helper method to streams - `streams.out_is_terminal()` that encapsulates this exact check.

I know that projects in the rust ecosystem like to move very fast wrt MSRV, and like to include a lot of dependencies, but that's not to my taste and I consider it to be one of the annoyances of the ecosystem.

Reply:
22d92b8f39893588be5d5fff6d7ee3f9abae6af2 should be the answer here.